spec/gold_miner/slack/client_spec.rb

```diff
@@ -45,131 +45,138 @@
+           "messages" => {
+             "matches" => [msg1, msg2],
+             "paging" => {"pages" => 1}
+           }}
         )
-        stub_slack_users_info_request(
-          token: token,
-          user_id: user1.id,
-          body: {"ok" => true, "user" => {"profile" => {"real_name" => user1.name}}}
-        )
-        stub_slack_users_info_request(
-          token: token,
-          user_id: user2.id,
-          body: {"ok" => true, "user" => {"profile" => {"real_name" => user2.name}}}
-        )
-        stub_slack_users_info_request(
-          token: token,
-          user_id: user3.id,
-          body: {"ok" => true, "user" => {"profile" => {"real_name" => user3.name}}}
-        )
-        stub_slack_users_info_request(
-          token: token,
-          user_id: "user4-id",
-          body: {"ok" => true, "user" => {"profile" => {"real_name" => "User 4"}}}
-        )
-        author_config = GoldMiner::AuthorConfig.new({
-          user1.username => {"link" => user1.link},
-          user2.username => {"link" => user2.link},
-          user3.username => {"link" => user3.link}
-        })
-        slack = GoldMiner::Slack::Client.build(api_token: token, author_config:).value!
-
-        messages = slack.search_interesting_messages_in("dev")
-
-        expect(messages).to match_array [
-          GoldMiner::Slack::Message.new(text: "TIL", author: user1, permalink: "https:///message-1-permalink.com"),
-          GoldMiner::Slack::Message.new(text: "Ruby tip/TIL: Array#sample...", author: user2, permalink: "https:///message-2-permalink.com"),
-          GoldMiner::Slack::Message.new(text: "Ruby tip: have fun!", author: user2, permalink: "https:///message-3-permalink.com"),
-          GoldMiner::Slack::Message.new(text: "CSS clamp() is so cool!", author: user3, permalink: "https:///message-4-permalink.com")
-        ]
-      end
+      }
+
+      user_info_query_time = 1
+      allow(sleepy_slack_client).to receive(:users_info).with({user: user1.id}) {
+        sleep(user_info_query_time)
+        deep_open_struct({"ok" => true, "user" => {"profile" => {"real_name" => user1.name}}})
+      }
+      allow(sleepy_slack_client).to receive(:users_info).with({user: user2.id}) {
+        sleep(user_info_query_time)
+        deep_open_struct({"ok" => true, "user" => {"profile" => {"real_name" => user2.name}}})
+      }
+
+      slack = described_class.build(api_token: "valid-token", slack_client: sleepy_slack_class).value!
+
+      t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      slack.search_messages(search_query)
+      total_elapsed_time = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0
+
+      # We have to sum the search and user info query times because they run
+      # sequentially. Both user info queries run in parallel, so we don't have
+      # to sum them twice.
+      overhead = 0.1
+      expect(total_elapsed_time).to be_within(overhead).of(search_query_time + user_info_query_time)
```

Comment:
Nice! I had no idea how to test this

Reply:
minor: It might be worth it to assert the following:
```rb
expect(sleepy_slack_client).to have_received(:search_messages).with({query: search_query}).once
expect(sleepy_slack_client).to have_received(:users_info).with({user: user1.id}).once
expect(sleepy_slack_client).to have_received(:users_info).with({user: user2.id}).once
```

I usually add those assertions to make sure the stub/mocks have actually been called.

In other words, without those assertions, this implementation of `search_messages` passes the test :D

```rb
module GoldMiner
  class Slack::Client
    # ...

    def search_messages(query)
      sleep 1.5
    end
```